### DISCUSSION TITLE
allowing duplicate package dependencies?

### DISCUSSION THREAD

**Discussion by wu-lee** [bundlewrap/bundlewrap#718] [Q&A] [answered]
New to bundlewrap here. Refugee from Ansible, which I've found to be the least-worst tool for the job so far (having tried Puppet, Chef, and researched a few others about a decade ago). I was getting tired of working around Ansible's clunky re-invention of imperative programming in YAML, and was looking for a stable python API for Ansible, but didn't find any. I am hopeful that bundlewrap is going to be better than the least-worst! So far so good, it's overall an improvement I think, just lacking in mindshare.

Anyway, as there's a bit of a shortage of bundlewrap online know-how, so I'm going to ask some questions. Here's my first.
 
I have two bundles, call them A and B. They both need git, so have a dependency on `pkg_apt:git` defined in `items.py`. 

However, I get an error when applying:  "bundlewrap.exceptions.BundleError: duplicate definition of pkg_apt:git". 

This would be fair enough if I didn't intend it - but is there a way of telling bundlewrap this is intentional, so not to worry about this and go ahead anyway?

**Comment by trehn** [comment]
Hi and welcome :)

This has come up quite a few times. So far, every item (such as `pkg_apt:git`) has to belong to exactly one bundle. Allowing multiple definitions of the same item would require bw to handle conflicting attributes (say one bundle wants `pkg_apt:git` installed, while another wants it uninstalled). It also creates ambiguity for things like `bw apply --skip bundle:A` - if an item is present in bundles A and B, should it be skipped or not?

So that's the design reason why this is the way it is.

The "proper" way to deal with the practical issue of multiple bundles wanting the same thing is through metadata. Create a dedicated `apt` bundle that has something like this in `items.py`:

```python
for pkg in node.metadata.get('apt/packages', set()):
    pkg_apt[pkg] = {}
```

In your other bundles, you can then do this in `metadata.py`:

```python
defaults = {
    'apt': {
        'packages': {
            'git',
        },
    },
}
```

While this is a little more complicated than just using `items.py` directly in each bundle, it has the added benefit that you can just add single packages to node metadata if you need it and don't need to write bundles that only install a single package.

Does that work for you?

**Comment by wu-lee** [comment]
Thanks, that works - but it sees to do this at the cost of a bit extra 
complexity, and a global impact on all the bundles which install 
packages.  Because of this problem with two bundles, I need to rewrite 
all my bundles?

Then there's PPAs - I think I'd have to factor all that into the apt 
bundle too.

Is there a well-written example of this apt bundle already existing 
which I can start from, which handles PPAs? I see there's one in 
kunsi/bundlewrap (link below), but it's not well documented and it isn't 
very obvious what it's doing and why. It seems to deal with other 
packaging systems, but looks a but hardwired to the package requirements 
of the author.

https://git.franzi.business/kunsi/bundlewrap/src/branch/main/bundles/apt

If this and apt bundle is a recommended solution, it would be good to 
know this sooner rather than later. The documentation could highlight 
this and point to a suggested implementation?

Thinking more: I do also wonder about your point about the conflicting 
attributes. The apt bundle solution solves the conflict at the cost of 
preventing *any* attributes being used (or at least, any difference).  
If I'm not using any different attributes on the various duplicate 
package requirements in the first place, then surely Bundlewrap has no 
problems to solve, and could just go ahead and perform the install.

Could Bundlewrap have a configuration which tells it to check for 
attribute conflicts when installing any given package, and only raise an 
error if there are any?

**Reply by trehn** [reply, marked as answer]
>The documentation could highlight this and point to a suggested implementation?

It really depends on how complex your repo is going to be. A dedicated apt bundle will not be necessary for everyone, so I don't want to push everyone down that path. At work we have a big repo with 1000+ nodes. The apt bundle there has 234 lines in `items.py`, 100 lines in `metadata.py` and 17 file templates. It's also quite opinionated, just like Kunsi's. A generic enough bundle that serves everone's preferences would be incredibly complex. That's why I believe it's better for everyone to come up with their own bundles that fit their individual needs.

>The apt bundle solution solves the conflict at the cost of preventing *any* attributes being used (or at least, any difference).

No, that was just a simplification I made. If you look at Kunsi's bundle, you'll notice they're allowing options to be passed. That is possible because at the metadata level, BundleWrap *does* have all the magic to allow for merging and reporting conflicts.

>Could Bundlewrap have a configuration which tells it to check for attribute conflicts when installing any given package, and only raise an error if there are any?

I don't see the point of a config option for that. The question is whether we can break up the 1:1 relationship between items and bundles and allow non-conflicting definitions of items. This was previously asked in #396 and I have now added that to the 5.0 milestone (since it would be a backwards-incompatible change). People keep running into this, so BundleWrap needs to do better there. The underlying architectural changes are significant though, so this will take a while.
